component/file_cache/file_cache_async.go

```diff
@@ -0,0 +1,331 @@
+	err := fc.NextComponent().DeleteDir(options)
+	if err != nil {
+		log.Err("FileCache::DeleteDir : %s failed", options.Name)
+		log.Err("FileCache::DeleteDir : Error is %s", err.Error())
+		// There is a chance that meta file for directory was not created in which case
+		// rest api delete will fail while we still need to cleanup the local cache for the same
+		return err
+	}
+	return nil
+}
+
+func (fc *FileCache) asyncCreateDir(options internal.CreateDirOptions) error {
+	err := fc.NextComponent().CreateDir(options)
+	if err != nil {
+		log.Err("FileCache::asyncCreateDir : %s failed", options.Name)
+		log.Err("FileCache::asyncDeleteDir : Error is %s", err.Error())
+		// There is a chance that meta file for directory was not created in which case
+		// rest api delete will fail while we still need to cleanup the local cache for the same
+		return err
+	}
+	return nil
+}
+
+func (fc *FileCache) asyncRenameDir(options internal.RenameDirOptions) error {
+
+	err := fc.NextComponent().RenameDir(options)
+	if err != nil {
+		log.Err("FileCache::RenameDir : error %s [%s]", options.Src, err.Error())
+		return err
+	}
+	return nil
+}
+
+func (fc *FileCache) asyncCreateFile(options internal.CreateFileOptions) error {
+
+	newF, err := fc.NextComponent().CreateFile(options)
+	if err != nil {
+		log.Err("FileCache::CreateFile : Failed to create file %s", options.Name)
+		return err
+	}
+	newF.GetFileObject().Close()
+	return nil
+}
+
+func (fc *FileCache) asyncChmod(options internal.ChmodOptions) error {
+
+	err := fc.NextComponent().Chmod(options)
+	err = fc.validateStorageError(options.Name, err, "Chmod", false)
+	if err != nil {
+		if err != syscall.EIO {
+			log.Err("FileCache::Chmod : %s failed to change mode [%s]", options.Name, err.Error())
+		} else {
+			fc.missedChmodList.LoadOrStore(options.Name, true)
+		}
+		return err
+	}
+	return nil
+}
+
+func (fc *FileCache) asyncChmodAndFlush(options internal.ChmodOptions) error {
+
+	//need to first flushFile before Chmod to ensure file is in cloud
+
+	flushFilePath := FlushFileAbstraction{}
+	flushFilePath.Name = options.Name
+
+	// We can allow for empty files to be flushed here because this only gets called by flushFile, meaning the user wants to close the file
+	err := fc.asyncFlushFile(flushFilePath)
+	if err != nil {
+		log.Err("FileCache::Chmod : %s failed to flushFile [%s]", options.Name, err.Error())
+		return err
+	}
+
+	err = fc.NextComponent().Chmod(options)
+	err = fc.validateStorageError(options.Name, err, "Chmod", false)
+	if err != nil {
+		if err != syscall.EIO {
+			log.Err("FileCache::Chmod : %s failed to change mode [%s]", options.Name, err.Error())
+		} else {
+			fc.missedChmodList.LoadOrStore(options.Name, true)
+		}
+		return err
+	}
+	return nil
+}
+
+func (fc *FileCache) asyncChown(options internal.ChownOptions) error {
+
+	// flushFilePath := FlushFileAbstraction{}
+	// flushFilePath.Name = options.Name
+
+	// if !fc.createEmptyFile {
```

Comment:
Does this commented out section need to be implemented later. Like there could still be an issue with empty file?

Reply:
I think we resolved this already! Deleted the comments